File: ios/RNTrueconfReactSdk.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(RNTrueconfReactSdk, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(server, NSString)
RCT_EXPORT_VIEW_PROPERTY(isCameraOn, BOOL)
RCT_EXPORT_VIEW_PROPERTY(isMuted, BOOL)

RCT_EXPORT_VIEW_PROPERTY(onServerStatus, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onStateChanged, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLogin, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLogout, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAccept, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onInvite, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onReject, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onRejectTimeout, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onConferenceStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onConferenceEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUserStatusUpdate, RCTDirectEventBlock)

#pragma mark - JS exported methods
RCT_EXPORT_METHOD(
  initSdk:(nonnull NSNumber *)node
  isMuted:(nonnull BOOL *)isMuted
  isCameraOn:(nonnull BOOL *)isCameraOn
)
RCT_EXPORT_METHOD(
  stopSdk:(nonnull NSNumber *)node
)
RCT_EXPORT_METHOD(
  makeCall:(nonnull NSNumber *)node
  to:(nonnull NSString *)to
)
RCT_EXPORT_METHOD(
  hangup:(nonnull NSNumber *)node
  forAll:(nonnull BOOL *)forAll
)
RCT_EXPORT_METHOD(
  acceptCall:(nonnull NSNumber *)node
  accept:(nonnull BOOL *)confId
)
RCT_EXPORT_METHOD(
  joinConf:(nonnull NSNumber *)node
  confId:(nonnull NSString *)confId
)
RCT_EXPORT_METHOD(
  login:(nonnull NSNumber *)node
  userId:(nonnull NSString *)userId
  password:(nonnull NSString *)password
  encryptPassword:(nonnull BOOL *)encryptPassword
  enableAutoLogin:(nonnull BOOL *)enableAutoLogin
)
RCT_EXPORT_METHOD(
  logout:(nonnull NSNumber *)node
)

@end

File: ios/RNTrueconfReactSdk.swift
@objc(RNTrueconfReactSdk)
class RNTrueconfReactSdk: RCTViewManager {
    var trueConfView: TrueConfView?

    override func view() -> UIView! {
        trueConfView = TrueConfView()
        trueConfView!.initViews()
        return trueConfView
    }

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    func getComponent(node: NSNumber) -> TrueConfView {
        return self.bridge.uiManager.view(
            forReactTag: node
            ) as! TrueConfView
    }

    @objc
    func initSdk(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.initSdk()
        }
    }

    @objc
    func stopSdk(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.stopSdk()
        }
    }

    @objc
    func makeCall(_ node: NSNumber, to: NSString) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.makeCall(to: to)
        }
    }

    @objc
    func hangup(_ node: NSNumber, forAll: Bool) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.hangup(forAll: forAll)
        }
    }

    @objc
    func login(_ node: NSNumber, userId: NSString, password: NSString, encryptPassword: Bool, enableAutoLogin: Bool) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.login(userId: userId, password: password, encryptPassword: encryptPassword, enableAutoLogin: enableAutoLogin)
        }
    }

    @objc
    func logout(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.logout()
        }
    }

    @objc
    func acceptCall(_ node: NSNumber, accept: Bool) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.acceptCall(accept: accept)
        }
    }

    @objc
    func joinConf(_ node: NSNumber, confId: String) {
        DispatchQueue.main.async {
            let component = self.getComponent(node: node)
            component.joinConf(confId: confId)
        }
    }

}
